src/ucp/core/ucp_mm.c

```diff
@@ -518,6 +518,31 @@ static ucs_status_t ucp_memh_register_gva(ucp_context_h context, ucp_mem_h memh,
     return UCS_OK;
 }
 
+static int ucp_memh_sys_dev_reachable(ucp_md_index_t md_index,
+                                      ucs_sys_device_t mem_sys_dev,
+                                      ucp_sys_dev_map_t sys_dev_map)
+{
+    ucs_sys_device_t sys_dev;
+
+    if ((mem_sys_dev == UCS_SYS_DEVICE_ID_UNKNOWN) || (sys_dev_map == 0)) {
+        return 1;
+    }
+
+    /*
+     * If at least one sys_dev is not reachable, do not register on it
+     * as we cannot know in advance which device is going to be used.
+     */
+    ucs_for_each_bit(sys_dev, sys_dev_map) {
+        if (!ucs_topo_is_memory_reachable(sys_dev, mem_sys_dev)) {
+            ucs_trace("md[%d] skipped: sys_dev=%u cannot reach mem_sys_dev=%u",
```

Comment:
It looks like this log may be confusing, because it is not guaranteed that the result of the function will be used exactly this way. `ucs_trace("md[%d] skipped: system device[%u] is not reachable", md_index, mem_attr.sys_dev);` should be in `ucp_memh_register_internal`. And you can remove `ucp_md_index_t md_index` from parameters of this function then.